File: modules/reebbon-cbr/ios/ReebbonCbrModule.swift
import ExpoModulesCore
import Foundation

public final class ReebbonCbrModule: Module {
  private let extractionQueue = DispatchQueue(
    label: "com.bakerscript.reebbon.cbr",
    qos: .userInitiated
  )

  public func definition() -> ModuleDefinition {
    Name("ReebbonCbr")

    AsyncFunction("extract") {
      (sourceUri: String, extractionId: String) throws -> [String: Any] in
      let job = try self.ownedJob(extractionId)
      guard !FileManager.default.fileExists(atPath: job.path) else {
        throw self.codedError(
          "ERR_CBR_DESTINATION_EXISTS",
          "The native CBR import directory already exists"
        )
      }

      do {
        try FileManager.default.createDirectory(
          at: job,
          withIntermediateDirectories: true
        )
        let localSource = try self.acquireSource(
          sourceUri,
          destination: job.appendingPathComponent("source.cbr")
        )
        let destination = job.appendingPathComponent(
          "extracted",
          isDirectory: true
        )
        let nativeResult: [String: Any]
        do {
          nativeResult = try ReebbonCbrBridge.extractSourcePath(
            localSource.path,
            destinationPath: destination.path
          ) as? [String: Any] ?? [:]
        } catch let error as NSError {
          let code = error.userInfo["code"] as? String ?? "ERR_CBR_NATIVE"
          throw self.codedError(code, error.localizedDescription, cause: error)
        }
        var result = nativeResult
        try FileManager.default.removeItem(at: localSource)
        result["directoryUri"] = destination.absoluteString
        return result
      } catch {
        try? FileManager.default.removeItem(at: job)
        throw error
      }
    }.runOnQueue(extractionQueue)

    AsyncFunction("cleanup") { (extractionId: String) throws in
      let job = try self.ownedJob(extractionId)
      if FileManager.default.fileExists(atPath: job.path) {
        do {
          try FileManager.default.removeItem(at: job)
        } catch {
          throw self.codedError(
            "ERR_CBR_CLEANUP",
            "The temporary CBR import directory could not be removed",
            cause: error
          )
        }
      }
    }.runOnQueue(extractionQueue)
  }

  private func acquireSource(_ sourceUri: String, destination: URL) throws -> URL {
    let source: URL
    if let parsed = URL(string: sourceUri), parsed.isFileURL {
      source = parsed
    } else if !sourceUri.contains("://") {
      source = URL(fileURLWithPath: sourceUri)
    } else {
      throw codedError(
        "ERR_CBR_SOURCE_ACCESS",
        "The selected CBR URI scheme is not supported"
      )
    }

    let securityAccess = source.startAccessingSecurityScopedResource()
    defer {
      if securityAccess {
        source.stopAccessingSecurityScopedResource()
      }
    }

    do {
      let values = try source.resourceValues(forKeys: [.fileSizeKey])
      guard let fileSize = values.fileSize, fileSize >= 0 else {
        throw codedError(
          "ERR_CBR_SOURCE_ACCESS",
          "The selected CBR archive size is unavailable"
        )
      }
      if UInt64(fileSize) > ReebbonCbrBridge.maximumArchiveBytes() {
        throw codedError(
          "ERR_CBR_ARCHIVE_SIZE_LIMIT",
          "The CBR archive exceeds the native input size limit"
        )
      }
      try FileManager.default.copyItem(at: source, to: destination)
      return destination
    } catch let error as CodedError where error.code.hasPrefix("ERR_CBR_") {
      throw error
    } catch {
      throw codedError(
        "ERR_CBR_SOURCE_ACCESS",
        "The selected CBR archive could not be copied locally",
        cause: error
      )
    }
  }

  private func ownedJob(_ extractionId: String) throws -> URL {
    let range = extractionId.range(
      of: "^[A-Za-z0-9][A-Za-z0-9._-]*$",
      options: .regularExpression
    )
    guard range?.lowerBound == extractionId.startIndex,
          range?.upperBound == extractionId.endIndex else {
      throw codedError(
        "ERR_CBR_UNSAFE_CLEANUP",
        "The CBR extraction identifier is invalid"
      )
    }

    let root = extractionRoot.standardizedFileURL
    let job = root.appendingPathComponent(
      extractionId,
      isDirectory: true
    ).standardizedFileURL
    guard job.path.hasPrefix(root.path + "/") else {
      throw codedError(
        "ERR_CBR_UNSAFE_CLEANUP",
        "The CBR import path is outside its native temporary root"
      )
    }
    return job
  }

  private func codedError(
    _ code: String,
    _ message: String,
    cause: Error? = nil
  ) -> Exception {
    let error = Exception(
      name: "ReebbonCbrException",
      description: message,
      code: code
    )
    error.cause = cause
    return error
  }

  private var extractionRoot: URL {
    FileManager.default.temporaryDirectory.appendingPathComponent(
      "reebbon-cbr",
      isDirectory: true
    )
  }
}

File: modules/reebbon-cbr/android/src/main/java/expo/modules/reebboncbr/ReebbonCbrModule.kt
package expo.modules.reebboncbr

import android.content.Context
import android.net.Uri
import android.os.Debug
import android.util.Log
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.InputStream

class ReebbonCbrModule : Module() {
  private val context: Context
    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()

  init {
    System.loadLibrary("reebbon-cbr")
  }

  override fun definition() = ModuleDefinition {
    Name("ReebbonCbr")

    AsyncFunction("extract") Coroutine { sourceUri: String, extractionId: String ->
      withContext(Dispatchers.IO) {
        extractArchive(sourceUri, extractionId)
      }
    }

    AsyncFunction("cleanup") Coroutine { extractionId: String ->
      withContext(Dispatchers.IO) {
        cleanupJob(extractionId)
      }
    }
  }

  private fun extractArchive(sourceUri: String, extractionId: String): Map<String, Any> {
    val job = ownedJob(extractionId)
    if (job.exists() || !job.mkdirs()) {
      throw CodedException(
        "ERR_CBR_DESTINATION_EXISTS",
        "The native CBR import directory could not be created",
        null,
      )
    }

    try {
      recordMemoryCheckpoint("before-source-acquisition", extractionId)
      val localSource = acquireSource(sourceUri, File(job, "source.cbr"))
      recordMemoryCheckpoint(
        "source-acquired",
        extractionId,
        "sourceBytes=${localSource.length()}",
      )
      val destination = File(job, "extracted")
      recordMemoryCheckpoint("extraction-start", extractionId)
      val values = nativeCall {
        extractNative(localSource.absolutePath, destination.absolutePath)
      }
      recordMemoryCheckpoint(
        "extraction-complete",
        extractionId,
        "entryCount=${values[0]},fileCount=${values[1]},totalBytes=${values[2]}",
      )
      if (!localSource.delete()) {
        throw CodedException(
          "ERR_CBR_CLEANUP",
          "The temporary CBR source copy could not be removed",
          null,
        )
      }
      recordMemoryCheckpoint("source-copy-removed", extractionId)
      return mapOf(
        "directoryUri" to Uri.fromFile(destination).toString(),
        "entryCount" to values[0],
        "fileCount" to values[1],
        "totalBytes" to values[2],
        "solid" to (values[3] == 1L),
      )
    } catch (error: Throwable) {
      if (!job.deleteRecursively() && job.exists()) {
        throw CodedException(
          "ERR_CBR_CLEANUP",
          "The failed CBR import directory could not be removed",
          error,
        )
      }
      throw error
    }
  }

  private fun acquireSource(sourceUri: String, destination: File): File {
    if (sourceUri.isBlank()) {
      throw CodedException(
        "ERR_CBR_SOURCE_ACCESS",
        "A CBR source URI is required",
        null,
      )
    }
    val maximumBytes = maxArchiveBytesNative()
    try {
      openSourceStream(sourceUri).use { input ->
        FileOutputStream(destination).use { output ->
          val buffer = ByteArray(SOURCE_COPY_BUFFER_BYTES)
          var totalBytes = 0L
          while (true) {
            val read = input.read(buffer)
            if (read < 0) {
              break
            }
            if (read == 0) {
              continue
            }
            val readBytes = read.toLong()
            if (totalBytes > maximumBytes - readBytes) {
              throw CodedException(
                "ERR_CBR_ARCHIVE_SIZE_LIMIT",
                "The CBR archive exceeds the native input size limit",
                null,
              )
            }
            output.write(buffer, 0, read)
            totalBytes += readBytes
          }
        }
      }
    } catch (error: CodedException) {
      throw error
    } catch (error: Throwable) {
      throw CodedException(
        "ERR_CBR_SOURCE_ACCESS",
        "The selected CBR archive could not be copied locally",
        error,
      )
    }
    return destination
  }

  private fun openSourceStream(sourceUri: String): InputStream {
    val uri = Uri.parse(sourceUri)
    return when (uri.scheme?.lowercase()) {
      "content" -> context.contentResolver.openInputStream(uri)
        ?: throw CodedException(
          "ERR_CBR_SOURCE_ACCESS",
          "The selected CBR content URI could not be opened",
          null,
        )
      "file" -> FileInputStream(
        uri.path ?: throw CodedException(
          "ERR_CBR_SOURCE_ACCESS",
          "The selected CBR file URI has no path",
          null,
        ),
      )
      null -> FileInputStream(sourceUri)
      else -> throw CodedException(
        "ERR_CBR_SOURCE_ACCESS",
        "The selected CBR URI scheme is not supported",
        null,
      )
    }
  }

  private fun cleanupJob(extractionId: String) {
    val job = ownedJob(extractionId)
    if (!job.deleteRecursively() && job.exists()) {
      throw CodedException(
        "ERR_CBR_CLEANUP",
        "The temporary CBR import directory could not be removed",
        null,
      )
    }
  }

  private fun ownedJob(extractionId: String): File {
    if (!EXTRACTION_ID.matches(extractionId)) {
      throw CodedException(
        "ERR_CBR_UNSAFE_CLEANUP",
        "The CBR extraction identifier is invalid",
        null,
      )
    }
    val root = File(context.cacheDir, TEMPORARY_ROOT).canonicalFile
    val job = File(root, extractionId).canonicalFile
    if (!job.path.startsWith(root.path + File.separator)) {
      throw CodedException(
        "ERR_CBR_UNSAFE_CLEANUP",
        "The CBR import path is outside its native temporary root",
        null,
      )
    }
    return job
  }

  private fun recordMemoryCheckpoint(
    stage: String,
    extractionId: String,
    details: String = "",
  ) {
    val memory = Debug.MemoryInfo()
    Debug.getMemoryInfo(memory)
    Log.i(
      MEMORY_LOG_TAG,
      "stage=$stage,extractionId=$extractionId,totalPssKb=${memory.totalPss}," +
        "nativePrivateDirtyKb=${memory.nativePrivateDirty},$details",
    )
  }

  private external fun extractNative(sourcePath: String, destinationPath: String): LongArray
  private external fun maxArchiveBytesNative(): Long

  private fun <T> nativeCall(block: () -> T): T = try {
    block()
  } catch (error: IllegalStateException) {
    val (code, message) = error.message
      ?.split('|', limit = 2)
      ?.let { parts -> parts.first() to parts.getOrElse(1) { parts.first() } }
      ?: ("ERR_CBR_NATIVE" to "The native CBR extractor failed")
    throw CodedException(code, message, error)
  }

  private companion object {
    const val MEMORY_LOG_TAG = "ReebbonCbrMemory"
    const val SOURCE_COPY_BUFFER_BYTES = 64 * 1024
    const val TEMPORARY_ROOT = "reebbon-cbr"
    val EXTRACTION_ID = Regex("^[A-Za-z0-9][A-Za-z0-9._-]*$")
  }
}
